### PR TITLE
Elasticsearch - refactor `_search_documents`

### PR DESCRIPTION
**Pull request by anakin87**
The method `_search_documents` was created to properly manage calls to the Elasticsearch client's search method and handle pagination. It also avoids code duplication.

**Changes** 
- improve this method, making it consider the `top_k` value if present (to stop iterating)
- introduce it in `_bm25_retrieval`
- add tests for retrieval with pagination, where this method plays a key role

### CODE REVIEW

File: document_stores/elasticsearch/src/elasticsearch_haystack/document_store.py [hunk 1]
```diff
@@ -115,8 +119,12 @@ def _search_documents(self, **kwargs) -> List[Document]:
                 from_=from_,
                 **kwargs,
             )
+
             documents.extend(self._deserialize_document(hit) for hit in res["hits"]["hits"])
-            from_ = len(documents)
+            from_ += len(documents)
```
Comment (silvanocerza) [inline document_stores/elasticsearch/src/elasticsearch_haystack/document_store.py:124]: This doesn't seem correct to me. 🤔

File: document_stores/elasticsearch/src/elasticsearch_haystack/document_store.py [hunk 2]
```diff
@@ -115,8 +119,12 @@ def _search_documents(self, **kwargs) -> List[Document]:
...
+
+            if top_k and from_ >= top_k:
```
Comment (silvanocerza) [inline document_stores/elasticsearch/src/elasticsearch_haystack/document_store.py:126]: Let's be explicit like we do above.

```suggestion
            if top_k is not None and from_ >= top_k:
```

File: document_stores/elasticsearch/src/elasticsearch_haystack/document_store.py [hunk 3]
```diff
@@ -326,14 +334,17 @@ def _bm25_retrieval(
...
-        return docs
+        if not scale_score:
+            return documents
+
+        documents_w_scaled_scores = []
+        for doc in documents:
+            doc.score = float(1 / (1 + np.exp(-np.asarray(doc.score / BM25_SCALING_FACTOR))))
+            documents_w_scaled_scores.append(doc)
+
+        return documents_w_scaled_scores
```
Comment (silvanocerza) [inline document_stores/elasticsearch/src/elasticsearch_haystack/document_store.py:347]: Can't we do this instead? There shouldn't be any need to create a new list. 🤔 

```suggestion
        if scale_score:
            for doc in documents:
                doc.score = float(1 / (1 + np.exp(-np.asarray(doc.score / BM25_SCALING_FACTOR))))

        return documents
```

File: document_stores/elasticsearch/tests/test_document_store.py [hunk 4]
```diff
@@ -1,6 +1,10 @@
 # SPDX-FileCopyrightText: 2023-present Silvano Cerza <silvanocerza@gmail.com>
 #
 # SPDX-License-Identifier: Apache-2.0
+
+# ruff: noqa: S311
```
Comment (silvanocerza) [inline document_stores/elasticsearch/tests/test_document_store.py:5]: Can we move this ignore to the line that actually uses `random`?

File: document_stores/elasticsearch/src/elasticsearch_haystack/document_store.py (hunk 1 above)
Comment (anakin87) [reply document_stores/elasticsearch/src/elasticsearch_haystack/document_store.py]: You're right!

File: document_stores/elasticsearch/tests/test_document_store.py (hunk 4 above)
Comment (anakin87) [reply document_stores/elasticsearch/tests/test_document_store.py]: I had some trouble doing it, but now I have managed.

**Review by silvanocerza — APPROVED** [review]
Go go! 🚀

**Pull request merged**